homeassistant/components/subaru/__init__.py

```diff
@@ -0,0 +1,250 @@
+                for component in SUPPORTED_PLATFORMS
+            ]
+        )
+    )
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+    if len(hass.data[DOMAIN]) == 0:
+        hass.data.pop(DOMAIN)
+
+    return unload_ok
+
+
+async def update_listener(hass, config_entry):
+    """Update when config_entry options update."""
+    controller = hass.data[DOMAIN][config_entry.entry_id]["controller"]
+    controller.set_update_interval(config_entry.options.get(CONF_HARD_POLL_INTERVAL))
+    controller.set_fetch_interval(config_entry.options.get(CONF_SCAN_INTERVAL))
+
+
+class SubaruEntity(Entity):
+    """Representation of a Subaru Device."""
+
+    def __init__(self, vehicle_info, coordinator):
+        """Initialize the Subaru Device."""
+        self.coordinator = coordinator
+        self.car_name = vehicle_info["display_name"]
+        self.vin = vehicle_info["vin"]
+        self._attributes = {}
+        self._should_poll = False
+        self.title = "entity"
+
+    @property
+    def name(self):
+        """Return name."""
+        return f"{self.car_name} {self.title}"
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return slugify(
+            "Subaru Model {} {} {}".format(
+                str(self.vin[3]).upper(), self.vin[-6:], self.title
+            )
+        )
+
+    @property
+    def icon(self):
+        """Return the icon of the sensor."""
+        return ICONS.get(self.title)
+
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        return self._should_poll
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the device."""
+        return self._attributes
+
+    @property
+    def device_info(self):
+        """Return the device_info of the device."""
+        return {
+            "identifiers": {(DOMAIN, self.vin)},
+            "name": self.car_name,
+            "manufacturer": "Subaru",
+        }
+
+    @property
+    def available(self):
+        """Return if entity is available."""
+        return self.coordinator.last_update_success
+
+    async def async_added_to_hass(self):
+        """Register state update callback."""
+        self.async_on_remove(
+            self.coordinator.async_add_listener(self.async_write_ha_state)
+        )
+
+    async def async_will_remove_from_hass(self):
+        """Prepare for unload."""
+
+    async def async_update(self):
+        """Update the state of the device."""
+        await self.coordinator.async_request_refresh()
+
+
+async def subaru_update(vehicle_info, controller):
+    """
+    Fetch or Update data, depending on how long it has been.
+
+    Subaru API calls assume a server side vehicle context
+    Data fetch/update must be done for each vehicle
+    """
+    data = {}
+
+    for vin in vehicle_info.keys():
+        # Only g2 api vehicles have data updates
+        if vehicle_info[vin]["api_gen"] == "g2":
```

Comment:
Maybe just move to next vehicle if not g2? Will be easier to follow the logic.
I would also refactor this into a function

Reply:
Agreed.  Changes made in 934d31e